com.unity.multiplayer.mlapi/Runtime/Core/SnapshotSystem.cs

```diff
@@ -0,0 +1,412 @@
+                if (m_NetworkManager.IsServer)
+                {
+                    for (int i = 0; i < m_NetworkManager.ConnectedClientsList.Count; i++)
+                    {
+                        var clientId = m_NetworkManager.ConnectedClientsList[i].ClientId;
+                        SendSnapshot(clientId);
+                    }
+                }
+                else
+                {
+                    SendSnapshot(m_NetworkManager.ServerClientId);
+                }
+
+                DebugDisplayStore(m_Snapshot, "Entries");
+                DebugDisplayStore(m_ReceivedSnapshot, "Received Entries");
+            }
+        }
+
+        // todo --M1--
+        // for now, the full snapshot is always sent
+        // this will change significantly
+        /// <summary>
+        /// Send the snapshot to a specific client
+        /// </summary>
+        /// <param name="clientId">The client index to send to</param>
+        private void SendSnapshot(ulong clientId)
+        {
+            // Send the entry index and the buffer where the variables are serialized
+            using (var buffer = PooledNetworkBuffer.Get())
+            {
+                WriteIndex(buffer);
+                WriteBuffer(buffer);
+
+                m_NetworkManager.MessageSender.Send(clientId, NetworkConstants.SNAPSHOT_DATA,
+                    NetworkChannel.SnapshotExchange, buffer);
+                buffer.Dispose();
+            }
+        }
+
+        /// <summary>
+        /// Write the snapshot index to a buffer
+        /// </summary>
+        /// <param name="buffer">The buffer to write the index to</param>
+        private void WriteIndex(NetworkBuffer buffer)
+        {
+            using (var writer = PooledNetworkWriter.Get(buffer))
+            {
+                writer.WriteInt16((short)m_Snapshot.LastEntry);
+                for (var i = 0; i < m_Snapshot.LastEntry; i++)
+                {
+                    WriteEntry(writer, in m_Snapshot.Entries[i]);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Read the snapshot index from a buffer
+        /// Stores the entry. Allocates memory if needed. The actual buffer will be read later
+        /// </summary>
+        /// <param name="reader">The reader to read the index from</param>
+        private void ReadIndex(NetworkReader reader)
+        {
+            Entry entry;
+            short entries = reader.ReadInt16();
+
+            for (var i = 0; i < entries; i++)
+            {
+                entry = ReadEntry(reader);
+                entry.Fresh = true;
+
+                int pos = m_ReceivedSnapshot.Find(entry.Key);
+                if (pos == Entry.NotFound)
+                {
+                    pos = m_ReceivedSnapshot.AddEntry(entry.Key.NetworkObjectId, entry.Key.BehaviourIndex,
+                        entry.Key.VariableIndex);
+                }
+
+                if (m_ReceivedSnapshot.Entries[pos].Length < entry.Length)
```

Comment:
Just to validate my understanding, this would happen if an entry already allocated grows in size is that it? Like if I have a NetworkList for example? (I'd document this, it wasn't clear on first read)

Reply:
Yes. I'll throw in a comment to make it more explicit.